### PR TITLE
[dashboard] use HTTP endpoint for telemetry

### PR DESCRIPTION
**Pull request by akosyakov**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at b953942</samp>

This pull request improves the analytics system of the Gitpod platform by using new server endpoints, refactoring the client and server code, adding new event types and properties, and enforcing rate limits. It also removes some deprecated code and dependencies. The main files affected are `Analytics.tsx`, `Modal.tsx`, `UserOnboarding.tsx`, `service.tsx`, `StartWorkspace.tsx`, `server.ts`, `gitpod-server-impl.ts`, `analytics-controller.ts`, and `rate-limitter.ts`.

</details>

## Related Issue(s)

Fixes EXP-914

#### Preview status

https://ak-dashboa44277b88ae.preview.gitpod-dev.com/workspaces

If preview env is down run following command from a Gitpod workspace for this PR:
```
export GITPOD_ANALYTICS="log"
time leeway run dev:preview --dont-test
```

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [x] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/dashboard/src/Analytics.tsx [hunk 1]
```diff
@@ -162,18 +221,30 @@ export const trackLocation = async (includePII: boolean) => {
...
-        includePII: includePII,
-        anonymousId: getAnonymousId(),
-        properties: props,
+    fetch("/_analytics/trackLocation", {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+        },
+        body: JSON.stringify({
+            //if the user is authenticated, let server determine the id. else, pass anonymousId explicitly.
```
Comment (geropl) [inline components/dashboard/src/Analytics.tsx:230]: Is this truy? if we use fetch, wasn't there an option we have to set to make it send the cookie?

File: components/server/src/analytics-controller.ts [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                this.trackLocation(req.user?.id, event, clientHeaderFields);
+                res.sendStatus(200);
+            } catch (e) {
+                console.error("failed to track location", e);
+                res.sendStatus(500);
+            }
+        });
+        router.post("/identifyUser", async (req, res) => {
+            try {
+                if (!req.user?.id) {
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply components/server/src/analytics-controller.ts:63]: I don't think we want to use it for any other clients. They should use proxy endpoint directly if they can.

[the remark this answers is not in the document]
Comment (geropl) [reply components/server/src/analytics-controller.ts:63]: :100:

File: components/server/src/analytics-controller.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+} from "@gitpod/gitpod-protocol/lib/analytics";
+import { createCookielessId, maskIp } from "./analytics";
+import { ClientHeaderFields, toClientHeaderFields } from "./express-util";
+import { Config } from "./config";
+import { RateLimited } from "./api/rate-limited";
+import { RateLimitter } from "./rate-limitter";
+import { RateLimiterRes } from "rate-limiter-flexible";
+
+@injectable()
+export class AnalyticsController {
```
Comment (akosyakov) [inline components/server/src/analytics-controller.ts:23]: BrowserAnalyticsController - be specific that it is only for our browser clients

File: components/server/src/analytics-controller.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                }
+                if (await this.rateLimitted(req.user?.id, "identifyUser", res)) {
+                    return;
+                }
+                const clientHeaderFields = toClientHeaderFields(req);
+                const event = req.body as RemoteIdentifyMessage;
+                this.identifyUser(req.user.id, event, clientHeaderFields);
+                res.sendStatus(200);
+            } catch (e) {
+                console.error("failed to track location", e);
```
Comment (mustard-mh) [inline components/server/src/analytics-controller.ts:74]: identify user

File: components/server/src/analytics-controller.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            try {
+                if (!req.user?.id) {
+                    res.sendStatus(401);
+                    return;
+                }
+                if (await this.rateLimitted(req.user?.id, "identifyUser", res)) {
+                    return;
+                }
+                const clientHeaderFields = toClientHeaderFields(req);
+                const event = req.body as RemoteIdentifyMessage;
```
Comment (mustard-mh) [inline components/server/src/analytics-controller.ts:70]: Should we validate body? or ignore them all since we have catched all errors with 500

**Comment by mustard-mh** [comment]
preview is down, going to run it again

File: components/server/src/analytics-controller.ts (hunk 5 above)
Comment (akosyakov) [reply components/server/src/analytics-controller.ts:70]: yeah, 500

**Comment by mustard-mh** [comment]
Maybe because it's debug level loggings, let me check

**Comment by mustard-mh** [comment]
Going to update LOG_LEVEL env of server manually. 
image

**Comment by mustard-mh** [comment]
After edit server deployment to set env `GITPOD_ANALYTICS_WRITER=log`  and `LOG_LEVEL=debug`, I can see loggings for telemetry. There's three loggings, properties are redacted, but it's working at least.

Do we expect `identifyUser` to delete `anonymousId` when `userId` is set?

```
{"component":"server","severity":"DEBUG","time":"2023-12-12T11:54:41.529Z","context":{"requestId":"8415f21c-4e26-44e6-ba3e-f09bd1ffc970","requestKind":"http","requestMethod":"/_analytics/identifyUser","contextTimeMs":1.344958999659866},"message":"analytics identify","payload":{"userId":"4c7d6390-0eda-4c78-8b37-04d8b73ecaba","anonymousId":"unidentified-user","traits":"[redacted:nested:object}]"}}
{"component":"server","severity":"DEBUG","time":"2023-12-12T11:54:15.218Z","context":{"requestId":"e0b009c2-0d1b-4d4e-bb10-44fc972f0436","requestKind":"http","requestMethod":"/_analytics/trackLocation","contextTimeMs":1.3095800001174212},"message":"analytics page","payload":{"userId":"4c7d6390-0eda-4c78-8b37-04d8b73ecaba","anonymousId":"unidentified-user","context":"[redacted:nested:object}]","properties":"[redacted:nested:object}]"}}
{"component":"server","severity":"DEBUG","time":"2023-12-12T11:55:19.979Z","context":{"requestId":"d287fabd-e0aa-4c48-b8a7-7780c1f4e912","requestKind":"http","requestMethod":"/_analytics/trackEvent","contextTimeMs":0.4317789999768138},"message":"analytics track","payload":{"userId":"4c7d6390-0eda-4c78-8b37-04d8b73ecaba","anonymousId":"unidentified-user","event":"path_changed","properties":"[redacted:nested:object}]"}}
```

**Review by mustard-mh — APPROVED** [review]
LGTM

**Pull request merged**